### DISCUSSION TITLE
New null handling operator

### DISCUSSION THREAD

**Discussion by LarryWMSN** [dotnet/csharplang#5169] [Language Ideas] [answered]
# New null handling operator feature
* [x]  Proposed
* [ ]  Prototype: Not Started
* [ ]  Implementation: Not Started
* [ ]  Specification: Not Started

## Summary
Return null if expression evaluates to null or then evaluate right side expression

Instead of:
```
double? foo = null;
var bar = foo is null ? (decimal?)null : Convert.ToDecimal(foo);

or

double? foo = null;
decimal? bar = null;
if (foo is not null)
{
   bar = ConvertToDecimal(foo);
}

```

Do this:
```
double? foo = null;
var bar = foo !? ConvertToDecimal(foo);
```

or 

Instead of:
```
List<Person> foo = null;
var bar = foo is null ? null : JsonSerializer.Serialize(foo);
```

Do this:
```
List<Person> foo = null;
var bar = foo !? JsonSerializer.Serialize(foo);
```

## Motivation
Easier syntax and prevents having to check for null inside each method or getting an unwanted result from a method when a null is passed in.

## Detailed design
Expression would return a nullable object of the result type of the right-side expression.

## Alternatives
```
?!
??!
!??
```

**Comment by theunrepentantgeek** [comment, marked as answer]
Duplicate of #4424. 

Based on a quick skim of that discussion, there seems to be little support for a new operator as extension methods can easily do the same job (I may have missed some details though - I'd encourage you to read that discussion carefully.)

**Reply by LarryWMSN** [reply]
It appears that proposal is the same as mine (I tried many searches before I created this proposal but didn't come across it). 

The discussion does present many workarounds including extension methods, but all of those do not cover what is being proposed which would work in all situations. Most of the discussion centers around the workarounds and where they could break, and the remaining discusses possible operator syntax. I my opinion the extension methods require more code or are harder to read than what is proposed

In general, there will always be little support for new operators without a compelling reason to do so, and I agree. But since there is really no suitable alternative that covers that proposal (or mine) I hope that someone follows up from the language team because I do think this is a compelling use case for a new operator.
